---
title: "`scmap` tutorial"
author: "Vladimir Kiselev, Andrew Yiu and Martin Hemberg"
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
vignette: >
  %\VignetteIndexEntry{`scmap` tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE, cache=TRUE}
library(googleVis)
op <- options(gvis.plot.tag='chart')
```

# Datasets

In this tutorial we will run `scmap` on the four human pancreas datasets, `xin`, `segerstolpe`, `muraro` and `baron`, which are used as positive controls. In the `segerstolpe` dataset we will remove cells labeled as __not applicable__ since it is unclear how to interpret this label and how it should be matched to the other datasets. In the `xin` dataset cells labeled as __alpha.contaminated__, __beta.contaminated__, __gamma.contaminated__ and __delta.contaminated__ were also removed since they likely correspond to cells of lower quality. All datasets in Bioconductor [`SingleCellExperiment`](http://bioconductor.org/packages/SingleCellExperiment) class format can be downloaded from our [website](https://hemberg-lab.github.io/scRNA.seq.datasets). The datasets can also be found in the `~/data` folder. Let's load the data:
```{r message=FALSE, warning=FALSE}
library(SingleCellExperiment)

# xin
xin <- readRDS("~/data/xin.rds")
xin <- xin[,colData(xin)$cell_type1 != "alpha.contaminated"]
xin <- xin[,colData(xin)$cell_type1 != "beta.contaminated"]
xin <- xin[,colData(xin)$cell_type1 != "delta.contaminated"]
xin <- xin[,colData(xin)$cell_type1 != "gamma.contaminated"]

# segerstolpe
segerstolpe <- readRDS("~/data/segerstolpe.rds")
segerstolpe <- segerstolpe[,colData(segerstolpe)$cell_type1 != "not applicable"]

# muraro
muraro <- readRDS("~/data/muraro.rds")

# baron
baron <- readRDS("~/data/baron-human.rds")
```

Overview of the datasets:
```{r}
xin
segerstolpe
muraro
```

By default we put the cell labels provided in the original publication into the `cell_type1` column of each dataset:
```{r}
as.character(unique(xin$cell_type1))
as.character(unique(segerstolpe$cell_type1))
as.character(unique(muraro$cell_type1))
```

In the following chapters we will be projecting `baron` dataset to the others using both `scmap-cluster` and `scmap-cell` methods (Fig. 1a).

# Feature selection

Now we will load `scmap` and for all of the reference datasets select the most informative features (genes) using the dropout feature selection method (Fig. S1a):
```{r fig.height=6}
library(scmap)
xin <- selectFeatures(xin, suppress_plot = FALSE)
segerstolpe <- selectFeatures(segerstolpe, suppress_plot = FALSE)
muraro <- selectFeatures(muraro, suppress_plot = FALSE)
```

Features are stored in the `scmap_features` column of the `rowData` slot of each dataset. By default `scmap` selects 500 features (it can also be controlled by setting `n_features` parameter):
```{r}
table(rowData(xin)$scmap_features)
table(rowData(segerstolpe)$scmap_features)
table(rowData(muraro)$scmap_features)
```

# scmap-cluster

## Index

The `scmap-cluster` index of a reference dataset is created by finding the median gene expression for each cluster. By default `scmap` uses the `cell_type1` column of the `colData` slot in the reference to identify clusters. Other columns can be manually selected by adjusting `cluster_col` parameter:
```{r}
xin <- indexCluster(xin)
segerstolpe <- indexCluster(segerstolpe)
muraro <- indexCluster(muraro)
```

The function `indexCluster` automatically writes the `scmap_cluster_index` item of the `metadata` slot of the reference dataset. The index can be visualized as a heatmap:
```{r fig.height=8}
library(pheatmap)
pheatmap(metadata(xin)$scmap_cluster_index, show_rownames = FALSE)
pheatmap(metadata(segerstolpe)$scmap_cluster_index, show_rownames = FALSE)
pheatmap(metadata(muraro)$scmap_cluster_index, show_rownames = FALSE)
```

Visual inspection suggests that the alpha, beta, gamma and delta cells are relatively similar compared to the other cell-types in the human pancreas.

## Projection

Once the `scmap-cluster` indexes have been generated we can use them to project the `baron` dataset. This can be done with one index at a time, but `scmap` allows for simultaneous projection to multiple indexes if they are provided as a list:
```{r message=FALSE, warning=FALSE}
scmapCluster_results <- scmapCluster(
  projection = baron, 
  index_list = list(
    muraro = metadata(muraro)$scmap_cluster_index,
    xin = metadata(xin)$scmap_cluster_index,
    segerstolpe = metadata(segerstolpe)$scmap_cluster_index
  )
)
```

## Results

`scmap-cluster` projects the query dataset to all projections defined in the `index_list`. The results of cell label assignements are merged into one matrix:
```{r}
head(scmapCluster_results$scmap_cluster_labs)
```

Corresponding similarities are stored in the `scmap_cluster_siml` item:
```{r}
head(scmapCluster_results$scmap_cluster_siml)
```

`scmap` also provides combined results of all reference dataset (choose labels corresponding to the largest similarity across reference datasets), as shown in Fig. 2d:
```{r}
head(scmapCluster_results$combined_labs)
```

## Visualisation

The results of `scmap-cluster` can be visualized as a Sankey diagram to show how cell-types are matched (`getSankey()` function). Note that the Sankey diagram will only be informative if both the query and the reference datasets have been clustered, but it is not necessary to have meaningful labels assigned to the query (`cluster1`, `cluster2` etc. is sufficient). `scmap` allows to visualise both reference-specific and combined results (Fig. 2cd):


```{r results='asis', tidy=FALSE}
plot(
  getSankey(
    colData(baron)$cell_type1, 
    scmapCluster_results$scmap_cluster_labs[,"xin"],
    plot_height = 400
  )
)
```


```{r results='asis', tidy=FALSE}
plot(
  getSankey(
    colData(baron)$cell_type1, 
    scmapCluster_results$combined_labs,
    plot_height = 400
  )
)
```

Clearly the combination strategy (projection to large number of references) improves `scmap` results.

## Quantification

If the cluster names match, as is the case for the pancreas datasets, then we can calculate the quality of the projection using Cohen's kappa. We can use the `kappa2` function from the `irr` package to do this:
```{r}
library(irr)
kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCluster_results$scmap_cluster_labs[,"xin"]
  )[scmapCluster_results$scmap_cluster_labs[,"xin"] != "unassigned", ]
)$value
```

Since kappa does not take the unassigned cells into consideration, this fraction also needs to be calculated to evaluate the projection:

```{r}
length(which(scmapCluster_results$scmap_cluster_labs[,"xin"] != "unassigned"))/
  ncol(baron)*100
```

We repeat these calculations for `muraro` and `segerstolpe`:
```{r}
kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCluster_results$scmap_cluster_labs[,"segerstolpe"]
  )[scmapCluster_results$scmap_cluster_labs[,"segerstolpe"] != "unassigned", ]
)$value

length(which(scmapCluster_results$scmap_cluster_labs[,"segerstolpe"] != "unassigned"))/
  ncol(baron)*100

kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCluster_results$scmap_cluster_labs[,"muraro"]
  )[scmapCluster_results$scmap_cluster_labs[,"muraro"] != "unassigned", ]
)$value

length(which(scmapCluster_results$scmap_cluster_labs[,"muraro"] != "unassigned"))/
  ncol(baron)*100
```

We note that all three projections are good with kappa values of $.8$ or greater. However, the fraction of unassigned cells is high, in particular for `xin`. However, as shown in a Sankey diagram above, using the combination strategy the results can be improved:
```{r}
kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCluster_results$combined_labs
  )[scmapCluster_results$combined_labs != "unassigned", ]
)$value

length(which(scmapCluster_results$combined_labs != "unassigned"))/
  ncol(baron)*100
```

## Properties

Next, we ask what properties of the datasets determine the quality of the projection. We start by comparing the number of expressed features (per cell):

```{r fig.height=6}
n_features_xin <- colSums(logcounts(xin) != 0)
n_features_muraro <- colSums(logcounts(muraro) != 0)
n_features_segerstolpe <- colSums(logcounts(segerstolpe) != 0)
n_features_baron <- colSums(logcounts(baron) != 0)
boxplot(
  list(
    xin = n_features_xin, 
    segerstolpe = n_features_segerstolpe, 
    muraro = n_features_muraro, 
    baron = n_features_baron
  )
)
```

Clearly, `baron` has far fewer genes detected than the other three datasets. In particular, we note that the difference in number of features matches the kappa value - the smallest difference is for `muraro`, which also has the smallest number of unassigned cells and the highest kappa. We hypothesize that the discrepancy in the number of features between the projection dataset (`baron`) and reference datasets (`xin`, `segerstolpe` and `muraro`) is making it challenging for `scmap-cluster` to find the correct projections. Indeed, it's also in agreement with experimental protocols used to obtain these datasets: droplet-based (`inDrop`) protocol for `baron` dataset versus full-length transcript methods used for `xin`, `segerstolpe` and `muraro` (`SMARTer`, `Smart-Seq2` and `CEL-Seq2` correspondingly). Interestingly, the reversed projections of full-length transcript datasets to `baron` were very good and did not show any technology-based biases (Table S2). An important implication of this result is that it will most likely work well to build a reference with shallowly sequenced cells and then later carry out in-depth studies with deeper sequencing.

# scmap-cell

## Stochasticity

`scmap-cell` contains k-means step which makes it stochastic, i.e. running it multiple times will provide slightly different results. Therefore, we will fix a random seed, so that a user will be able to exactly reproduce our results:

```{r}
set.seed(1)
```


## Index

In the `scmap-cell` index is created by a product quantiser algorithm in a way that every cell in the reference is identified with a set of sub-centroids found via k-means clustering based on a subset of the features.

```{r message=FALSE, warning=FALSE}
xin <- indexCell(xin)
segerstolpe <- indexCell(segerstolpe)
muraro <- indexCell(muraro)
```

Unlike `scmap-cluster` index `scmap-cell` index contains information about each cell and therefore can not be easily visualised. `scmap-cell` index consists of two items:

```{r}
names(metadata(muraro)$scmap_cell_index)
```

### Sub-centroids

`subcentroids` contains coordinates of subcentroids of low dimensional subspaces defined by selected features, `k` and `M` parameters of the product quantiser algorithm (see `?indexCell`). 

For the `muraro` dataset:

* `muraro` dataset contains $N = 2126$ cells
* We selected $f = 500$ features (`scmap` default)
* `M` was calculated as $f / 10 = 50$ (`scmap` default for $f \le 1000$). `M` is the number of low dimensional subspaces
* Number of features in any low dimensional subspace equals to $f / M = 10$
* `k` was calculated as $k = \sqrt{N} \approx 46$ (`scmap` default). 

```{r}
length(metadata(muraro)$scmap_cell_index$subcentroids)
dim(metadata(muraro)$scmap_cell_index$subcentroids[[1]])
metadata(muraro)$scmap_cell_index$subcentroids[[1]][,1:5]
```


### Sub-clusters

For each of the `M` partitions of the feature space, `subclusters` contains information about which cluster the cells belong to:

```{r}
dim(metadata(muraro)$scmap_cell_index$subclusters)
metadata(muraro)$scmap_cell_index$subclusters[1:5,1:5]
```

## Projection

Once the `scmap-cell` indexes have been generated we can use them to project the `baron` dataset. This can be done with one index at a time, but as before `scmap` allows for simultaneous projection to multiple indexes if they are provided as a list:

```{r}
scmapCell_results <- scmapCell(
  baron, 
  list(
    xin = metadata(xin)$scmap_cell_index,
    segerstolpe = metadata(segerstolpe)$scmap_cell_index,
    muraro = metadata(muraro)$scmap_cell_index
  )
)
```

## Results

`scmapCell_results` contains results of projection for each reference dataset in a list:
```{r}
names(scmapCell_results)
```

For each dataset there are two matricies. The `cells` matrix contains the top 10 (`scmap` default) cell IDs of the cells of the reference dataset that a given cell of the projection dataset is closest to:

```{r}
scmapCell_results$xin$cells[,1:3]
```

The `similarities` matrix contains corresponding cosine similarities:

```{r}
scmapCell_results$xin$similarities[,1:3]
```

## Cluster annotation

If cell cluster annotation is available for the reference datasets, in addition to finding top 10 nearest neighbours `scmap-cell` can also annotate the cells from the projection dataset using the labels of the reference. It does so by looking at the top 3 nearest neighbours (`scmap` default) and if they all belong to the same cluster in the reference and their maximum similarity is higher than a threshold ($0.5$ is the `scmap` default), then a projection cell is assigned to the corresponding reference cluster:

```{r}
scmapCell_clusters <- scmapCell2Cluster(
  baron, 
  scmapCell_results, 
  list(
    colData(xin)$cell_type1,
    colData(segerstolpe)$cell_type1,
    colData(muraro)$cell_type1
  )
)
```

`scmap-cell` results are in the same format as the ones provided by `scmap-cluster` (see above):

```{r}
head(scmapCell_clusters$scmap_cluster_labs)
```

The corresponding similarities are stored in the `scmap_cluster_siml` item:
```{r}
head(scmapCell_clusters$scmap_cluster_siml)
```

```{r}
head(scmapCell_clusters$combined_labs)
```

## Visualisation

If the query cells are annotated with cluster names, then we can visualize the projection using Sankey diagrams as before:

```{r results='asis', tidy=FALSE}
plot(
  getSankey(
    colData(baron)$cell_type1, 
    scmapCell_clusters$scmap_cluster_labs[,"xin"],
    plot_height = 400
  )
)
```

```{r results='asis', tidy=FALSE}
plot(
  getSankey(
    colData(baron)$cell_type1, 
    scmapCell_clusters$combined_labs,
    plot_height = 400
  )
)
```

## Quantification

Similarly, we can quantify the accuracy of the projection by calculating Cohen's kappa and fraction of unassigned cells:

```{r}
kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCell_clusters$scmap_cluster_labs[,"xin"]
  )[scmapCell_clusters$scmap_cluster_labs[,"xin"] != "unassigned", ]
)$value

length(which(scmapCell_clusters$scmap_cluster_labs[,"xin"] != "unassigned"))/
  ncol(baron)*100

kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCell_clusters$scmap_cluster_labs[,"segerstolpe"]
  )[scmapCell_clusters$scmap_cluster_labs[,"segerstolpe"] != "unassigned", ]
)$value

length(which(scmapCell_clusters$scmap_cluster_labs[,"segerstolpe"] != "unassigned"))/
  ncol(baron)*100

kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCell_clusters$scmap_cluster_labs[,"muraro"]
  )[scmapCell_clusters$scmap_cluster_labs[,"muraro"] != "unassigned", ]
)$value

length(which(scmapCell_clusters$scmap_cluster_labs[,"muraro"] != "unassigned"))/
  ncol(baron)*100

kappa2(
  data.frame(
    colData(baron)$cell_type1,
    scmapCell_clusters$combined_labs
  )[scmapCell_clusters$combined_labs != "unassigned", ]
)$value

length(which(scmapCell_clusters$combined_labs != "unassigned"))/
  ncol(baron)*100
```

## Properties

Comparing the results with the ones for `scmap-cluster`, we conclude that the kappa values are similar to before, while the fraction of unassigned cells is much lower. Thus, for this example, we conclude that `scmap-cell` performs better.

# sessionInfo()

```{r, echo=FALSE}
sessionInfo()
```
